pkg/operator/controller/garden/garden/components.go

```diff
@@ -867,6 +894,30 @@ func getAPIServerDomains(domains []operatorv1alpha1.DNSDomain) []operatorv1alpha
 	return apiServerDomains
 }
 
+// GetAPIServerSNIDomains returns the domains which match a SNI domain pattern.
+func GetAPIServerSNIDomains(domains []string, sni operatorv1alpha1.SNI) []string {
+	var sniDomains []string
+
+	for _, domainPattern := range sni.DomainPatterns {
+		if strings.HasPrefix(domainPattern, "*.") {
+			patternWithoutWildcard := domainPattern[1:]
+			for _, domain := range domains {
+				domainWithoutSuffix := strings.TrimSuffix(domain, patternWithoutWildcard)
+				if strings.HasSuffix(domain, patternWithoutWildcard) && len(domainWithoutSuffix) > 0 && !strings.Contains(domainWithoutSuffix, ".") {
+					sniDomains = append(sniDomains, domain)
+				}
+			}
+			continue
+		}
+
+		if slices.Contains(domains, domainPattern) {
+			sniDomains = append(sniDomains, domainPattern)
+		}
+	}
+
+	return sniDomains
+}
```

Comment:
I have trouble understanding what is going on here. What is the expected input and what should happen here? Maybe a few comments highlighting what domains should be handled by each case could help.

Reply:
I added a few comments and split one if statement into two. Hope, it is easier to understand now 😄